### DISCUSSION TITLE
Best practice for dynamic routing

### DISCUSSION THREAD

**Discussion by gamegee** [vercel/next.js#19129] [Help] [answered]
Hey guys.

I'm new to nextjs and I'm getting used to react-router (v5+) where you can define routes dynamically.
So you can easily compose routes with pages/modules, and I have no idea what is the best practice to do the same in nextjs.

Let's take a basic example, I have 4 pages:
- `/sign-in`
Contains simple form to authenticate

- `/app`
Contains Navbar and app home page

- `/app/settings/user`
Contains Navbar, settings sidebar and user settings form

- `/app/settings/account`
Contains Navbar, settings sidebar and account settings form

In react router we can implement the routing like so:
```tsx
// MainRouter.tsx
<Route path="/sign-in" component={SignInPage} />
<Route path="/app" component={AppPages} />
```

```tsx
// AppPages.tsx
<Navbar>
<Route path="/app" component={Dashboard} />
<Route path="/app/settings" component={SettingsPages} />
```

```tsx
// SettingsPages.tsx
<SettingsSidebar>
<Route exact path="/app/settings">
  <Redirect to="/app/settings/user" />
</Route>
<Route path="/app/settings/user" component={UserSettings} />
<Route path="/app/settings/account" component={AccountSettings} />
```
---
The main benefit is you only instanciate your `<Navbar>` and `<SettingsSidebar>` once in the code.

With nextJS, it seems that only the `_app.tsx` can be used to wrap route, and I have no idea what would be the best practice to handle such cases.

**Comment by radoslawgrochowski** [comment, marked as answer]
https://adamwathan.me/2019/10/17/persistent-layout-patterns-in-nextjs/ ?

**Reply by gamegee** [reply]
Thx ! Seems to be the cleaner pattern I've seen so far, but it is still lot of overhead for such simple feature.

I would love to see dynamic routing supported as first-class feature in nextjs too :)
